anndata/_io/merge.py

```diff
@@ -0,0 +1,647 @@
+from functools import singledispatch
+from .specs import read_elem, write_elem
+from .._core.merge import (
+    _resolve_dim,
+    StrategiesLiteral,
+    intersect_keys,
+    unify_dtypes,
+    merge_indices,
+    merge_dataframes,
+    resolve_merge_strategy,
+    gen_reindexer,
+    gen_inner_reindexers,
+    reduce,
+    concat_arrays,
+    MissingVal,
+    Reindexer,
+)
+from .._core.sparse_dataset import SparseDataset
+from .._core.file_backing import to_memory
+from ..experimental import read_dispatched
+import pandas as pd
+from typing import (
+    Any,
+    Callable,
+    Collection,
+    Optional,
+    Sequence,
+    Union,
+    Literal,
+    MutableMapping,
+    Mapping,
+)
+import typing
+from pathlib import Path
+import numpy as np
+
+from ..compat import ZarrGroup, ZarrArray, H5Group, H5Array
+
+from scipy.sparse import csr_matrix, csc_matrix
+
+MAX_LOAD_SIZE = 60_000_000
+
+
+# TODO: Not good
+class IdentityReindexer:
+    def __init__(self):
+        self.no_change = True
+
+    def __call__(self, x, *args, **kwargs):
+        return x
+
+
+def _df_index(df: Union[ZarrGroup, H5Group]) -> pd.Index:
+    index_key = df.attrs["_index"]
+    return pd.Index(read_elem(df[index_key]))
+
+
+def _requires_reindexing(indices) -> bool:
+    init_elem = indices[0]
+    return any(not np.array_equal(init_elem, elem) for elem in indices[1:])
+
+
+def write_concat_mappings(
+    mappings,
+    output_group: Union[ZarrGroup, H5Group],
+    keys,
+    path,
+    axis=0,
+    index=None,
+    reindexers=None,
+    fill_value=None,
+):
+    """
+    Write a list of mappings to a zarr group.
+    """
+    mapping_group = output_group.create_group(path)
+    mapping_group.attrs.update(
+        {
+            "encoding-type": "dict",
+            "encoding-version": "0.1.0",
+        }
+    )
+    for k in keys:
+        elems = [m[k] for m in mappings]
+        write_concat_sequence(
+            elems,
+            output_group=mapping_group,
+            out_path=k,
+            axis=axis,
+            index=index,
+            reindexers=reindexers,
+            fill_value=fill_value,
+        )
+
+
+SPARSE_MATRIX = {"csc_matrix", "csr_matrix"}
+
+EAGER_TYPES = {"dataframe", "awkward-array"}
+
+
+def read_group(group: Union[ZarrGroup, H5Group]):
+    """Read the group until
+    SparseDataset, Array or EAGER_TYPES are encountered."""
```

Comment:
Could this be called `read_as_backed`? I think `read_group` is a bit too generic.